src/utils.rs

```diff
@@ -0,0 +1,161 @@
+//! Basic utils
+use crate::errors::NovaError;
+use ff::PrimeField;
+
+pub fn vector_add<F: PrimeField>(a: &Vec<F>, b: &Vec<F>) -> Result<Vec<F>, NovaError> {
+  if a.len() != b.len() {
+    return Err(NovaError::InvalidIndex);
+  }
+
+  let mut res = Vec::with_capacity(a.len());
+  for i in 0..a.len() {
+    res.push(a[i] + b[i]);
+  }
+
+  Ok(res)
+}
+
+pub fn vector_elem_product<F: PrimeField>(a: &Vec<F>, e: &F) -> Result<Vec<F>, NovaError> {
+  let mut res = Vec::with_capacity(a.len());
+  for i in 0..a.len() {
+    res.push(a[i] * e);
+  }
+
+  Ok(res)
+}
+
+#[allow(dead_code)]
+pub fn matrix_vector_product<F: PrimeField>(
+  matrix: &Vec<Vec<F>>,
+  vector: &Vec<F>,
+) -> Result<Vec<F>, NovaError> {
+  if matrix.len() == 0 || matrix[0].len() == 0 {
+    return Err(NovaError::InvalidIndex);
+  }
+
+  if matrix[0].len() != vector.len() {
+    return Err(NovaError::InvalidIndex);
+  }
+
+  let mut res = Vec::with_capacity(matrix.len());
+  for i in 0..matrix.len() {
+    let mut sum = F::ZERO;
+    for j in 0..matrix[i].len() {
+      sum += matrix[i][j] * vector[j];
+    }
+    res.push(sum);
+  }
+
+  Ok(res)
+}
+
+// Matrix vector product where matrix is sparse
+// First element is row index, second column, third value stored
+pub fn matrix_vector_product_sparse<F: PrimeField>(
+  matrix: &Vec<(usize, usize, F)>,
+  vector: &Vec<F>,
+) -> Result<Vec<F>, NovaError> {
+  if matrix.len() == 0 {
+    return Err(NovaError::InvalidIndex);
+  }
+
+  // Find the maximum row index in the matrix
+  let max_row = matrix.iter().map(|r| r.0).max().unwrap() + 1;
+  if max_row > vector.len() {
+    return Err(NovaError::InvalidIndex);
+  }
+
+  let mut res = vec![F::ZERO; max_row];
+  for &(row, col, value) in matrix {
+    res[row] += value * vector[col];
+  }
+
+  Ok(res)
+}
+
+pub fn hadamard_product<F: PrimeField>(a: &Vec<F>, b: &Vec<F>) -> Result<Vec<F>, NovaError> {
+  if a.len() != b.len() {
+    return Err(NovaError::InvalidIndex);
+  }
+
+  let mut res = Vec::with_capacity(a.len());
+  for i in 0..a.len() {
+    res.push(a[i] * b[i]);
+  }
+
+  Ok(res)
+}
+
+#[allow(dead_code)]
+pub fn to_F_vec<F: PrimeField>(v: Vec<u64>) -> Vec<F> {
+  v.iter().map(|x| F::from(*x)).collect()
+}
+
+#[allow(dead_code)]
+pub fn to_F_matrix<F: PrimeField>(m: Vec<Vec<u64>>) -> Vec<Vec<F>> {
+  m.iter().map(|x| to_F_vec(x.clone())).collect()
+}
+
+#[allow(dead_code)]
+pub fn to_F_matrix_sparse<F: PrimeField>(m: Vec<(usize, usize, u64)>) -> Vec<(usize, usize, F)> {
+  m.iter().map(|x| (x.0, x.1, F::from(x.2))).collect()
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use pasta_curves::Fq;
```

Comment:
There were recent PRs (e.g., #171, #165) to make tests generic over the curves. I wonder if it makes sense to continue that approach to allow library users to switch curves and still run these tests more easily.

Reply:
Thanks, agree! Will adjust tests accordingly.

Reply:
Done for this module. We did it w.r.t. PrimeField for these types as that's all they need.

(Will update other tests similarly once consensus on approach, see https://github.com/privacy-scaling-explorations/Nova/issues/34)